solr/core/src/test/org/apache/solr/cloud/NestedShardedAtomicUpdateTest.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cloud;
+
+import java.io.IOException;
+import java.util.List;
+
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.common.SolrDocument;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.common.params.SolrParams;
+import org.junit.Test;
+
+public class NestedShardedAtomicUpdateTest extends AbstractFullDistribZkTestBase {
+
+  public NestedShardedAtomicUpdateTest() {
+    stress = 0;
+    sliceCount = 4;
+    schemaString = "sharded-block-schema.xml";
+  }
+
+  @Override
+  protected String getCloudSolrConfig() {
+    return "solrconfig-block-atomic-update.xml";
+  }
+
+  @Override
+  protected String getCloudSchemaFile() {
+    return "sharded-block-schema.xml";
+  }
+
+  @Test
+  @ShardsFixed(num = 4)
+  public void test() throws Exception {
+    boolean testFinished = false;
+    try {
+      doNestedInplaceUpdateTest();
+      doRootShardRoutingTest();
+      testFinished = true;
+    } finally {
+      if (!testFinished) {
+        printLayoutOnTearDown = true;
+      }
+    }
+  }
+
+  public void doRootShardRoutingTest() throws Exception {
+    assertEquals(4, cloudClient.getZkStateReader().getClusterState().getCollection(DEFAULT_COLLECTION).getSlices().size());
+    final String[] ids = {"c", "d", "e", "f"};
+
+    assertEquals("size of ids to index should be the same as the number of clients", clients.size(), ids.length);
+    // for now,  we know how ranges will be distributed to shards.
+    // may have to look it up in clusterstate if that assumption changes.
+
+    SolrInputDocument doc = sdoc("id", "a", "level_s", "root");
+
+    final SolrParams params = new ModifiableSolrParams().set("update.chain", "nested-rtg").set("wt", "json").set("_route_", "a");
+
+    int which = (params.get("_route_").hashCode() & 0x7fffffff) % clients.size();
+    SolrClient aClient = clients.get(which);
+
+    indexDoc(aClient, params, doc);
+
+    aClient.commit();
```

Comment:
again; some of these commits are superfluous and would better be randomized so exercise different internal code paths.